File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.py
class Solution:
    def searchRange(self, nums: List[int], target: int) -> List[int]:
        def firstOccurence(nums, target):
            lo = 0
            hi = len(nums)-1
            while lo <= hi:
                i = (lo+hi)//2
                if nums[i] == target:
                    # find first occurence
                    if i-1 >= 0 and nums[i-1] == target:
                        hi = i - 1
                    else:
                        return i
                elif nums[i] > target:
                    hi = i-1
                elif nums[i] < target:
                    lo = i+1
            return -1


        def lastOccurence(nums, target):
            lo = 0
            hi = len(nums)-1
            while lo <= hi:
                i = (lo+hi)//2
                if nums[i] == target:
                    # find last occurence
                    if i+1 <= hi and nums[i+1] == target:
                        lo = i + 1
                    else:
                        return i
                elif nums[i] > target:
                    hi = i-1
                elif nums[i] < target:
                    lo = i+1
            return -1

        return [firstOccurence(nums, target), lastOccurence(nums, target)]    
        

File: 34-find-first-and-last-position-of-element-in-sorted-array/34-find-first-and-last-position-of-element-in-sorted-array.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        return new int[] {firstOcc(nums,target), lastOccr(nums,target)};
    }
    
    public int firstOcc(int[] nums,int target){
        int start = 0;
        int end = nums.length-1;
        
        while(start<=end){
            int mid = start + (end-start)/2;
            if(target == nums[mid]){
                if(mid-1>=0 && nums[mid-1]==target)
                    end = mid-1;
                else
                    return mid;
            }
            else if(target > nums[mid])
                start = mid+1;
            else
                end = mid-1;
        }
        return -1;
    }
        
    public int lastOccr(int[] nums,int target){
        int start = 0;
        int end = nums.length-1;
        
        while(start<=end){
            int mid = start+(end-start)/2;
            if(target == nums[mid]){
                if(mid+1<nums.length && nums[mid+1]==target)
                    start = mid+1;
                else
                    return mid;
            }
            else if(target > nums[mid])
                start = mid+1;
            else
                end = mid-1;
        }
        return -1;
    }
}
